<a href="https://colab.research.google.com/github/PranayUmredkar97/EDA-Hotel-Bookings-Capstone-Project-1/blob/main/Rossmann_Sales_Prediction_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Sales Prediction : Predicting sales of a major store chain Rossmann</u></b>

## <b> Problem Description </b>

### Rossmann operates over 3,000 drug stores in 7 European countries. Currently, Rossmann store managers are tasked with predicting their daily sales for up to six weeks in advance. Store sales are influenced by many factors, including promotions, competition, school and state holidays, seasonality, and locality. With thousands of individual managers predicting sales based on their unique circumstances, the accuracy of results can be quite varied.

### You are provided with historical sales data for 1,115 Rossmann stores. The task is to forecast the "Sales" column for the test set. Note that some stores in the dataset were temporarily closed for refurbishment.

![](https://drive.google.com/uc?export=view&id=19oj6uW3GPQE0TxSrnLPQa9Y5h_GQ11VE)

# **Import Packages**

In [2]:
#imporgt required packages and libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [ ]:
#set style for graph
sns.set_style(style='darkgrid')

# Setting fontsizes,fontweight,label weight for lebels,titles.
plt.rcParams["font.weight"] = "bold"
plt.rcParams["axes.labelweight"] = "bold"
plt.rcParams["axes.titlesize"] = 25
plt.rcParams["axes.titleweight"] = 'bold'
plt.rcParams['xtick.labelsize']=15
plt.rcParams['ytick.labelsize']=15
plt.rcParams["axes.labelsize"] = 20
plt.rcParams["legend.fontsize"] = 15
plt.rcParams["legend.title_fontsize"] = 15

# **Data Preperation**

## **Mounting Drive**

In [3]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [4]:
#csv file location
path1 = '/content/drive/MyDrive/Machine Learning Regression/store.csv'
path2 = '/content/drive/MyDrive/Machine Learning Regression/Rossmann Stores Data.csv'

## **Reading data in notebook**

In [14]:
Stores = pd.read_csv(path1)
Sales = pd.read_csv(path2)

## **Data Inspection**

In [15]:
#first look checking top 5 rows of data
print('Sales data')
Sales.head()

Sales data


Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [16]:
print('Stores data')
Stores.head()

Stores data


Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [18]:
#merge dataset

raw_df = pd.merge(Sales, Stores , how ='left', on = 'Store' )
raw_df.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday StoreType Assortment  CompetitionDistance  \
0              1         c          a               1270.0   
1              1         a          a                570.0   
2              1         a          a              14130.0   
3              1         c          c                620.0   
4              1         a          a              29910.0   

   CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                        9.0                    2008.0       0   
1                       11.0                    2007.0       1   
2                       12.0                    2006.0       1   
3                        9.0                    2009.0       0   
4                        4.0                    2015.0       0   

   Promo2SinceWeek  Promo2SinceYear    PromoInterval  
0              NaN              NaN              NaN  
1             13.0           2010.0  Jan,Apr,Jul,Oct  
2             14.0           2011.0  Jan,Apr,Jul,Oct  
3              NaN              NaN              NaN  
4              NaN              NaN              NaN

In [19]:
#check information about data
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   Store                      1017209 non-null  int64  
 1   DayOfWeek                  1017209 non-null  int64  
 2   Date                       1017209 non-null  object 
 3   Sales                      1017209 non-null  int64  
 4   Customers                  1017209 non-null  int64  
 5   Open                       1017209 non-null  int64  
 6   Promo                      1017209 non-null  int64  
 7   StateHoliday               1017209 non-null  object 
 8   SchoolHoliday              1017209 non-null  int64  
 9   StoreType                  1017209 non-null  object 
 10  Assortment                 1017209 non-null  object 
 11  CompetitionDistance        1014567 non-null  float64
 12  CompetitionOpenSinceMonth  693861 non-null   float64
 13  CompetitionO



*   Dataset have 1017209 entries and total of 18 columns
*   6 columns contains null values



## <b> Data Description </b>

### <b>Rossmann Stores Data.csv </b> - historical data including Sales
### <b>store.csv </b> - supplemental information about the stores


### <b><u>Data fields</u></b>
### Most of the fields are self-explanatory. The following are descriptions for those that aren't.

* #### Id - an Id that represents a (Store, Date) duple within the test set
* #### Store - a unique Id for each store
* #### Sales - the turnover for any given day (this is what you are predicting)
* #### Customers - the number of customers on a given day
* #### Open - an indicator for whether the store was open: 0 = closed, 1 = open
* #### StateHoliday - indicates a state holiday. Normally all stores, with few exceptions, are closed on state holidays. Note that all schools are closed on public holidays and weekends. a = public holiday, b = Easter holiday, c = Christmas, 0 = None
* #### SchoolHoliday - indicates if the (Store, Date) was affected by the closure of public schools
* #### StoreType - differentiates between 4 different store models: a, b, c, d
* #### Assortment - describes an assortment level: a = basic, b = extra, c = extended
* #### CompetitionDistance - distance in meters to the nearest competitor store
* #### CompetitionOpenSince[Month/Year] - gives the approximate year and month of the time the nearest competitor was opened
* #### Promo - indicates whether a store is running a promo on that day
* #### Promo2 - Promo2 is a continuing and consecutive promotion for some stores: 0 = store is not participating, 1 = store is participating
* #### Promo2Since[Year/Week] - describes the year and calendar week when the store started participating in Promo2
* #### PromoInterval - describes the consecutive intervals Promo2 is started, naming the months the promotion is started anew. E.g. "Feb,May,Aug,Nov" means each round starts in February, May, August, November of any given year for that store

In [28]:
#check the summary statistics of data
raw_df.describe(include='all').T

count unique              top    freq  \
Store                      1017209.0    NaN              NaN     NaN   
DayOfWeek                  1017209.0    NaN              NaN     NaN   
Date                         1017209    942       2015-07-31    1115   
Sales                      1017209.0    NaN              NaN     NaN   
Customers                  1017209.0    NaN              NaN     NaN   
Open                       1017209.0    NaN              NaN     NaN   
Promo                      1017209.0    NaN              NaN     NaN   
StateHoliday                 1017209      5                0  855087   
SchoolHoliday              1017209.0    NaN              NaN     NaN   
StoreType                    1017209      4                a  551627   
Assortment                   1017209      3                a  537445   
CompetitionDistance        1014567.0    NaN              NaN     NaN   
CompetitionOpenSinceMonth   693861.0    NaN              NaN     NaN   
CompetitionOpenSinceYear    693861.0    NaN              NaN     NaN   
Promo2                     1017209.0    NaN              NaN     NaN   
Promo2SinceWeek             509178.0    NaN              NaN     NaN   
Promo2SinceYear             509178.0    NaN              NaN     NaN   
PromoInterval                 509178      3  Jan,Apr,Jul,Oct  293122   

                                  mean          std     min     25%     50%  \
Store                       558.429727   321.908651     1.0   280.0   558.0   
DayOfWeek                     3.998341     1.997391     1.0     2.0     4.0   
Date                               NaN          NaN     NaN     NaN     NaN   
Sales                      5773.818972  3849.926175     0.0  3727.0  5744.0   
Customers                   633.145946   464.411734     0.0   405.0   609.0   
Open                          0.830107     0.375539     0.0     1.0     1.0   
Promo                         0.381515     0.485759     0.0     0.0     0.0   
StateHoliday                       NaN          NaN     NaN     NaN     NaN   
SchoolHoliday                 0.178647     0.383056     0.0     0.0     0.0   
StoreType                          NaN          NaN     NaN     NaN     NaN   
Assortment                         NaN          NaN     NaN     NaN     NaN   
CompetitionDistance        5430.085652    7715.3237    20.0   710.0  2330.0   
CompetitionOpenSinceMonth     7.222866     3.211832     1.0     4.0     8.0   
CompetitionOpenSinceYear   2008.690228     5.992644  1900.0  2006.0  2010.0   
Promo2                        0.500564          0.5     0.0     0.0     1.0   
Promo2SinceWeek              23.269093    14.095973     1.0    13.0    22.0   
Promo2SinceYear            2011.752774      1.66287  2009.0  2011.0  2012.0   
PromoInterval                      NaN          NaN     NaN     NaN     NaN   

                              75%      max  
Store                       838.0   1115.0  
DayOfWeek                     6.0      7.0  
Date                          NaN      NaN  
Sales                      7856.0  41551.0  
Customers                   837.0   7388.0  
Open                          1.0      1.0  
Promo                         1.0      1.0  
StateHoliday                  NaN      NaN  
SchoolHoliday                 0.0      1.0  
StoreType                     NaN      NaN  
Assortment                    NaN      NaN  
CompetitionDistance        6890.0  75860.0  
CompetitionOpenSinceMonth    10.0     12.0  
CompetitionOpenSinceYear   2013.0   2015.0  
Promo2                        1.0      1.0  
Promo2SinceWeek              37.0     50.0  
Promo2SinceYear            2013.0   2015.0  
PromoInterval                 NaN      NaN

In [29]:
#Creating a copy of dataset before working with it so that we can have original data untouched
df = raw_df.copy()

In [30]:
#check the columns names
df.columns

Index(['Store', 'DayOfWeek', 'Date', 'Sales', 'Customers', 'Open', 'Promo',
       'StateHoliday', 'SchoolHoliday', 'StoreType', 'Assortment',
       'CompetitionDistance', 'CompetitionOpenSinceMonth',
       'CompetitionOpenSinceYear', 'Promo2', 'Promo2SinceWeek',
       'Promo2SinceYear', 'PromoInterval'],
      dtype='object')

In [32]:
#checking columns and their unique values
cat_columns = df.select_dtypes(include=['object'])\

#create for loop to print unique values in categorical columns
for col in cat_columns:
  print(f'Unique values in categorical columns {col} are: {cat_columns[col].unique()[:8]}\n')

Unique values in categorical columns Date are: ['2015-07-31' '2015-07-30' '2015-07-29' '2015-07-28' '2015-07-27'
 '2015-07-26' '2015-07-25' '2015-07-24']

Unique values in categorical columns StateHoliday are: ['0' 'a' 'b' 'c' 0]

Unique values in categorical columns StoreType are: ['c' 'a' 'd' 'b']

Unique values in categorical columns Assortment are: ['a' 'c' 'b']

Unique values in categorical columns PromoInterval are: [nan 'Jan,Apr,Jul,Oct' 'Feb,May,Aug,Nov' 'Mar,Jun,Sept,Dec']



In [36]:
num_columns= df._get_numeric_data()

for col in num_columns:
  print(f'Unique values in Numerical columns {col} are: {num_columns[col].unique()[:10]}\n')

Unique values in Numerical columns Store are: [ 1  2  3  4  5  6  7  8  9 10]

Unique values in Numerical columns DayOfWeek are: [5 4 3 2 1 7 6]

Unique values in Numerical columns Sales are: [ 5263  6064  8314 13995  4822  5651 15344  8492  8565  7185]

Unique values in Numerical columns Customers are: [ 555  625  821 1498  559  589 1414  833  687  681]

Unique values in Numerical columns Open are: [1 0]

Unique values in Numerical columns Promo are: [1 0]

Unique values in Numerical columns SchoolHoliday are: [1 0]

Unique values in Numerical columns CompetitionDistance are: [ 1270.   570. 14130.   620. 29910.   310. 24000.  7520.  2030.  3160.]

Unique values in Numerical columns CompetitionOpenSinceMonth are: [ 9. 11. 12.  4. 10.  8. nan  3.  6.  5.]

Unique values in Numerical columns CompetitionOpenSinceYear are: [2008. 2007. 2006. 2009. 2015. 2013. 2014. 2000. 2011.   nan]

Unique values in Numerical columns Promo2 are: [0 1]

Unique values in Numerical columns Promo2SinceWeek a

## **Handlin missing values**

In [47]:
#check null values 
df.isnull().sum().sort_values(ascending=False)[:8]


PromoInterval                508031
Promo2SinceYear              508031
Promo2SinceWeek              508031
CompetitionOpenSinceYear     323348
CompetitionOpenSinceMonth    323348
CompetitionDistance            2642
DayOfWeek                         0
Promo2                            0
dtype: int64

In [48]:
percentage_missing = df.isnull().sum()*100/len(df)
percentage_missing.sort_values(ascending =False)[:8]

PromoInterval                49.943620
Promo2SinceYear              49.943620
Promo2SinceWeek              49.943620
CompetitionOpenSinceYear     31.787764
CompetitionOpenSinceMonth    31.787764
CompetitionDistance           0.259730
DayOfWeek                     0.000000
Promo2                        0.000000
dtype: float64